src/Java.Interop.BootstrapTasks/Java.Interop.BootstrapTasks/JdkInfo.cs

```diff
@@ -14,74 +17,265 @@ public class JdkInfo : Task
+				return null;
+			Version v;
+			if (!Version.TryParse (m.Groups ["version"].Value, out v)) {
+				return null;
+			}
+			return v;
+		}
+
+		void WritePropertyFile (string jarPath, string javacPath, string jdkJvmPath, IEnumerable<string> includes)
+		{
+			var msbuild = XNamespace.Get ("http://schemas.microsoft.com/developer/msbuild/2003");
+			var project = new XElement (msbuild + "Project",
+				new XElement (msbuild + "Choose",
+					new XElement (msbuild + "When", new XAttribute ("Condition", " '$(JdkJvmPath)' == '' "),
+						new XElement (msbuild + "PropertyGroup",
+							new XElement (msbuild + "JdkJvmPath", jdkJvmPath)),
+						new XElement (msbuild + "ItemGroup",
+							includes.Select (i => new XElement (msbuild + "JdkIncludePath", new XAttribute ("Include", i)))))),
+				new XElement (msbuild + "PropertyGroup",
+					new XElement (msbuild + "JavaCPath", new XAttribute ("Condition", " '$(JavaCPath)' == '' "),
+						javacPath),
+					new XElement (msbuild + "JarPath", new XAttribute ("Condition", " '$(JarPath)' == '' "),
+						jarPath)));
+			project.Save (PropertyFile.ItemSpec);
+		}
+
+		void WriteMakeFragmentFile (string jarPath, string javacPath, string jdkJvmPath, IEnumerable<string> includes)
+		{
+			using (var o = new StreamWriter (MakeFragmentFile.ItemSpec)) {
+				o.WriteLine ($"JI_JAR_PATH          := {jarPath}");
+				o.WriteLine ($"JI_JAVAC_PATH        := {javacPath}");
+				o.WriteLine ($"JI_JDK_INCLUDE_PATHS := {string.Join (" ", includes)}");
+				o.WriteLine ($"JI_JVM_PATH          := {jdkJvmPath}");
+			}
+		}
+
+		string GetJavaHomePathFromEnvironment ()
+		{
+			var java_home = Environment.GetEnvironmentVariable ("JAVA_HOME");
+			if (!string.IsNullOrEmpty (java_home))
+				return java_home;
+			return null;
+		}
+
+		string GetJavaHomePathFromMachine (Version maxVersion)
+		{
+			var java_homes  = GetJavaHomePathsFromDirectory (JdksRoot)
+				.Concat (GetJavaHomePathsFromJava ())
+				.Concat (GetJavaHomePathsFromWindowsRegistry ())
+				.Distinct ()
+				.Where (d => Directory.Exists (d))
+				.Select (jh => new {
+					Path    = jh,
+					Version = GetVersionFromPath (jh),
+				})
+				.Where (v => maxVersion == null ? true : v.Version <= maxVersion)
+				.OrderByDescending (v => v.Version)
+				.Select (v => v.Path)
+				.ToList ();
+
+			foreach (var p in java_homes) {
+				Log.LogMessage (MessageImportance.Low, $"  Possible JAVA_HOME location: {p}");
+			}
+
+			return java_homes.FirstOrDefault ();
+		}
+
+		IEnumerable<string> GetJavaHomePathsFromDirectory (string jdksRoot)
+		{
+			if (string.IsNullOrEmpty (jdksRoot))
+				yield break;
+			if (!Directory.Exists (jdksRoot))
+				yield break;
+			foreach (var d in Directory.EnumerateDirectories (jdksRoot)) {
+				var h = d;
+				if (OnMacOS)
+					h = Path.Combine (h, "Contents", "Home");
+				yield return h;
+			}
+		}
+
+		IEnumerable<string> GetJavaHomePathsFromJava ()
+		{
+			var java  = Path.DirectorySeparatorChar == '/' ? "java" : "java.exe";
```

Comment:
Once again, this decision can be made in the task init - we use a finite set of binary names, they could be stored in variables in task init. Would save a small amount of time, but still would be savings.

Reply:
I would prefer an `OS.Platform` check here rather than checking the DirectorySeparatorChar.

Reply:
It's certainly more descriptive...